jerry-core/ecma/base/ecma-globals.h

```diff
@@ -279,151 +279,95 @@ typedef struct ecma_property_t
+ * Ecma object is built-in or lexical environment.
+ *   - built-in, if object type is less than ECMA_LEXICAL_ENVIRONMENT_TYPES_START
+ *   - lexical environment, if object type is greater or equal than ECMA_LEXICAL_ENVIRONMENT_TYPES_START
  */
-#define ECMA_OBJECT_GC_VISITED_POS (ECMA_OBJECT_GC_NEXT_CP_POS + \
-                                    ECMA_OBJECT_GC_NEXT_CP_WIDTH)
-#define ECMA_OBJECT_GC_VISITED_WIDTH (1)
-
-
-/* Objects' only part */
+#define ECMA_OBJECT_FLAG_BUILT_IN_OR_LEXICAL_ENV 0x08
 
 /**
- * Attribute 'Extensible'
+ * This object is visited by the garbage collector.
  */
-#define ECMA_OBJECT_OBJ_EXTENSIBLE_POS (ECMA_OBJECT_GC_VISITED_POS + \
-                                        ECMA_OBJECT_GC_VISITED_WIDTH)
-#define ECMA_OBJECT_OBJ_EXTENSIBLE_WIDTH (1)
+#define ECMA_OBJECT_FLAG_GC_VISITED 0x10
 
 /**
- * Implementation internal object type (ecma_object_type_t)
+ * Extensible object.
  */
-#define ECMA_OBJECT_OBJ_TYPE_POS (ECMA_OBJECT_OBJ_EXTENSIBLE_POS + \
-                                  ECMA_OBJECT_OBJ_EXTENSIBLE_WIDTH)
-#define ECMA_OBJECT_OBJ_TYPE_WIDTH (3)
+#define ECMA_OBJECT_FLAG_EXTENSIBLE 0x20
 
 /**
- * Compressed pointer to prototype object (ecma_object_t)
+ * Value for increasing or decreasing the object reference counter.
  */
-#define ECMA_OBJECT_OBJ_PROTOTYPE_OBJECT_CP_POS (ECMA_OBJECT_OBJ_TYPE_POS + \
-                                                 ECMA_OBJECT_OBJ_TYPE_WIDTH)
-#define ECMA_OBJECT_OBJ_PROTOTYPE_OBJECT_CP_WIDTH (ECMA_POINTER_FIELD_WIDTH)
+#define ECMA_OBJECT_REF_ONE (1u << 6)
 
 /**
- * Flag indicating whether the object is a built-in object
+ * Maximum value of the object reference counter (1023).
  */
-#define ECMA_OBJECT_OBJ_IS_BUILTIN_POS (ECMA_OBJECT_OBJ_PROTOTYPE_OBJECT_CP_POS + \
-                                        ECMA_OBJECT_OBJ_PROTOTYPE_OBJECT_CP_WIDTH)
-#define ECMA_OBJECT_OBJ_IS_BUILTIN_WIDTH (1)
+#define ECMA_OBJECT_MAX_REF (0x3ffu << 6)
 
 /**
- * Size of structure for objects
- */
-#define ECMA_OBJECT_OBJ_TYPE_SIZE (ECMA_OBJECT_OBJ_IS_BUILTIN_POS + \
-                                   ECMA_OBJECT_OBJ_IS_BUILTIN_WIDTH)
-
-
-/* Lexical environments' only part */
-
-/**
- * Type of lexical environment (ecma_lexical_environment_type_t).
- */
-#define ECMA_OBJECT_LEX_ENV_TYPE_POS (ECMA_OBJECT_GC_VISITED_POS + \
-                                        ECMA_OBJECT_GC_VISITED_WIDTH)
-#define ECMA_OBJECT_LEX_ENV_TYPE_WIDTH (1)
-
-/**
- * Compressed pointer to outer lexical environment
+ * Description of ECMA-object or lexical environment
+ * (depending on is_lexical_environment).
  */
-#define ECMA_OBJECT_LEX_ENV_OUTER_REFERENCE_CP_POS (ECMA_OBJECT_LEX_ENV_TYPE_POS + \
-                                                    ECMA_OBJECT_LEX_ENV_TYPE_WIDTH)
-#define ECMA_OBJECT_LEX_ENV_OUTER_REFERENCE_CP_WIDTH (ECMA_POINTER_FIELD_WIDTH)
+typedef struct ecma_object_t
+{
+  /** object type, flags, and refs */
+  uint16_t type_flags_refs;
```

Comment:
Could you add a more descriptive comment for this? For example it is not trivial why 
`if (object_p->type_flags_refs < ECMA_OBJECT_MAX_REF)` will work. (I know it is correct at the moment :) ) I think we need to give some help here. How should be tpye/flags/refs used?